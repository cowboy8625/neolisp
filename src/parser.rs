use crate::ast::{Expr, Span, Spanned};
use crate::error::{Error, ErrorType};
use chumsky::prelude::*;

pub fn parse_expr() -> impl Parser<char, Spanned<Expr>, Error = Error> {
    recursive(|expr| {
        let comment = just(';').then(take_until(just('\n'))).padded();

        let int = text::int(10);

        let number = int
            .or(just('-').repeated().then(int).map(|(sign, num)| {
                let sign = if sign.len() % 2 == 0 { "" } else { "-" };
                format!("{sign}{num}")
            }))
            .map_with_span(|s: String, span: Span| {
                Spanned::from((Expr::Number(s.parse().unwrap()), span))
            })
            .padded();

        let boolean = just("true")
            .map_with_span(|_, span| Spanned::from((Expr::Bool(true), span)))
            .or(just("false").map_with_span(|_, span| (Expr::Bool(false), span).into()))
            .padded();

        let punctuation = one_of(r#"!$,_-./:;?+<=>#%&*@[\]{|}`^~"#);
        let letters = one_of("abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ");
        let digit = one_of("0123456789");

        let symbol = choice((letters.clone(), punctuation.clone()))
            .then(choice((letters, punctuation, digit)).repeated())
            .padded()
            .map_with_span(|(start, end), span| {
                (
                    Expr::Symbol(format!("{start}{}", end.iter().collect::<String>())),
                    span,
                )
                    .into()
            });

        let string = just('"')
            .ignore_then(choice((
                none_of('"')
                    .repeated()
                    .at_least(1)
                    .collect::<String>()
                    .map(|s| s.replace("\\t", "\t"))
                    .map(|s| s.replace("\\n", "\n"))
                    .map_with_span(|s, span| (Expr::String(s), span).into())
                    .then_ignore(just('"')),
                just('"').map_with_span(|_, span| (Expr::String("".to_string()), span).into()),
            )))
            .padded();

        let atom = number
            .or(string)
            .or(boolean)
            .or(symbol)
            .or(expr.clone())
            .padded_by(comment.repeated());

        let quoted = just('\'')
            .ignore_then(atom.clone())
            .padded_by(comment.repeated())
            .map_with_span(|e: Spanned<Expr>, span: Span| {
                (
                    Expr::List(vec![
                        (Expr::Symbol("quote".to_string()), span.clone()).into(),
                        e,
                    ]),
                    span,
                )
                    .into()
            });

        let list = atom
            .repeated()
            .delimited_by(
                just('(').map_err(|e| Error::with_label(e, ErrorType::MissingOpeningParenthesis)),
                just(')').map_err(|e| Error::with_label(e, ErrorType::MissingClosingParenthesis)),
            )
            .padded_by(comment.repeated())
            .padded()
            .map_with_span(|list, span| Spanned::from((Expr::List(list), span)));

        list.or(quoted)
    })
}

pub fn parser() -> impl Parser<char, Vec<Spanned<Expr>>, Error = Error> {
    parse_expr().repeated().padded().then_ignore(end())
}
