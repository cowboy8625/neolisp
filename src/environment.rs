use crate::builtins::*;
use crate::parser::Expr;
use std::collections::HashMap;

#[derive(Debug, Clone)]
pub struct Env {
    pub data: HashMap<String, Expr>,
    pub outer: Option<Box<Env>>,
}

macro_rules! init_builtin {
    ($doc:ident, $symbol:expr, $name:ident) => {
        (
            $symbol.to_string(),
            crate::parser::Expr::Builtin($name, $doc.get($symbol).unwrap().to_string()),
        )
    };
    ($doc:ident, $name:ident) => {
        (
            stringify!($name).to_string(),
            Expr::Builtin($name, $doc.get(stringify!($name)).unwrap().to_string()),
        )
    };
}

impl Env {
    pub fn new() -> Self {
        let docs = load_doc();
        let data = HashMap::from([
            init_builtin!(docs, "+", add),
            init_builtin!(docs, "-", sub),
            init_builtin!(docs, "=", eq),
            init_builtin!(docs, ">", gt),
            init_builtin!(docs, "<", lt),
            init_builtin!(docs, ">=", gte),
            init_builtin!(docs, "<=", lte),
            init_builtin!(docs, "typeof", type_of), // typeof is a keyword in rust
            init_builtin!(docs, and),
            init_builtin!(docs, or),
            init_builtin!(docs, not),
            init_builtin!(docs, help),
            init_builtin!(docs, print),
            init_builtin!(docs, list),
            init_builtin!(docs, cons),
            init_builtin!(docs, car),
            init_builtin!(docs, cdr),
            init_builtin!(docs, append),
            init_builtin!(docs, reverse),
            init_builtin!(docs, nth),
            init_builtin!(docs, length),
            init_builtin!(docs, map),
            init_builtin!(docs, fold),
        ]);
        Self { data, outer: None }
    }
}
