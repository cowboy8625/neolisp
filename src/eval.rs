use crate::environment::Env;
use crate::parser::{Expr, Func, Lambda, Span, Spanned};

pub type EvalResult = Result<Spanned<Expr>, String>;

pub fn eval(expr: &Spanned<Expr>, env: &mut Env) -> EvalResult {
    match &expr.expr {
        Expr::Bool(_) => Ok(expr.clone()),
        Expr::Number(_) => Ok(expr.clone()),
        Expr::String(_) => Ok(expr.clone()),
        Expr::Symbol(symbol) => eval_symbol(symbol, env),
        Expr::List(list) => eval_list(list, env),
        _ => unreachable!("invalid expr: {expr:?}"),
    }
}

fn eval_symbol(symbol: &str, env: &mut Env) -> EvalResult {
    if let Some(outer) = &mut env.outer {
        outer
            .as_mut()
            .data
            .get(symbol)
            .cloned()
            .ok_or(format!("undefined symbol: {symbol}"))
    } else if env.data.contains_key(symbol) {
        env.data
            .get(symbol)
            .cloned()
            .ok_or(format!("undefined symbol: {symbol}"))
    } else {
        Err(format!("undefined symbol: {symbol}"))
    }
}

fn eval_list(expr: &[Spanned<Expr>], env: &mut Env) -> EvalResult {
    let Some(head) = expr.first() else {
        return Err("empty list are not allowed".to_string());
    };
    let args = &expr[1..];
    match &head.expr {
        Expr::Symbol(symbol) if symbol == "var" => eval_define_variable(args, env),
        Expr::Symbol(symbol) if symbol == "let" => eval_define_let(args, env),
        Expr::Symbol(symbol) if symbol == "if" => eval_if(args, env),
        Expr::Symbol(symbol) if symbol == "lambda" => eval_define_lambda(head.span.clone(), args),
        Expr::Symbol(symbol) if symbol == "fn" => eval_define_fn(head.span.clone(), args, env),
        Expr::Symbol(symbol) if symbol == "quote" => eval_quote(args),
        _ => {
            let evaled_head = eval(head, env)?;
            match evaled_head.expr {
                Expr::Builtin(builtin, _) => builtin(evaled_head.span, args, env),
                Expr::Lambda(func) => eval_lambda(evaled_head.span, &func, args, env),
                Expr::Func(func) => eval_func(&func, args, env),
                expr => Err(format!(
                    "{:?}: list head expected to be a function, but got {expr:?}",
                    evaled_head.span
                )),
            }
        }
    }
}

fn eval_define_fn(starting_span: Span, args: &[Spanned<Expr>], env: &mut Env) -> EvalResult {
    let Some(name_symbol) = args.first() else {
        return Err("missing name, params and body in fn definition".to_string());
    };

    let Some(params) = args.get(1) else {
        return Err("missing params and body in fn definition".to_string());
    };

    let mut body_index = 2;
    let help_doc = match args.get(2) {
        Some(maybe_doc) if matches!(maybe_doc.expr, Expr::String(_)) => {
            body_index += 1;
            Some(Box::new(maybe_doc.clone()))
        }
        _ => None,
    };

    let Some(body) = args.get(body_index) else {
        return Err("missing body in fn definition".to_string());
    };
    let Expr::Symbol(name) = name_symbol.expr.clone() else {
        return Err("fn name must be a symbol".to_string());
    };
    env.data.insert(
        name.clone(),
        (
            Expr::Func(Func {
                name: Box::new(name_symbol.clone()),
                params: Box::new(params.clone()),
                body: Box::new(body.clone()),
                help_doc,
            }),
            starting_span,
        )
            .into(),
    );
    Ok(name_symbol.clone())
}

fn eval_func(func: &Func, args: &[Spanned<Expr>], env: &mut Env) -> EvalResult {
    let mut new_env = env.clone();
    let Spanned {
        expr: Expr::List(params),
        ..
    } = &*func.params
    else {
        unreachable!("if this message is displayed, there is a bug in the in the eval function for params buts be a list");
    };

    for (param, arg) in params.iter().zip(args) {
        let Expr::Symbol(param) = &param.expr else {
            unreachable!("if this message is displayed, there is a bug in the in the eval function for params buts be a list of symbol");
        };
        let i = eval(arg, env)?;
        new_env.data.insert(param.to_string(), i);
    }
    eval(&func.body, &mut new_env)
}

fn eval_define_variable(args: &[Spanned<Expr>], env: &mut Env) -> EvalResult {
    if args.len() != 2 {
        return Err(format!(
            "invalid number of arguments, expected 2, got {}",
            args.len()
        ));
    }
    let Some(name) = args.first() else {
        return Err("empty list are not allowed in var definition".to_string());
    };
    let value = eval(&args[1], env)?;
    env.data.insert(name.to_string(), value);
    Ok(name.clone())
}

fn eval_define_let(args: &[Spanned<Expr>], env: &mut Env) -> EvalResult {
    if args.len() < 2 {
        return Err("let requires at least two arguments".to_string());
    }
    let Some(last) = args.last() else {
        return Err("empty list are not allowed in let definition".to_string());
    };
    let mut new_env = env.clone();
    for arg in &args[0..args.len() - 1] {
        let Spanned {
            expr: Expr::List(list),
            ..
        } = arg
        else {
            return Err("let arguments must be a list".to_string());
        };
        eval_define_variable(list, &mut new_env)?;
    }
    eval(last, &mut new_env)
}

fn eval_define_lambda(lambda_span: Span, args: &[Spanned<Expr>]) -> EvalResult {
    let Some(Spanned {
        expr: Expr::List(params),
        span: params_span,
    }) = args.first()
    else {
        return Err("lambda params must be a list".to_string());
    };
    let Some(Spanned {
        expr: Expr::List(body),
        span: body_span,
    }) = args.get(1)
    else {
        return Err("lambda body must be a list".to_string());
    };
    let lambda = Lambda {
        params: Box::new((Expr::List(params.clone()), params_span).into()),
        body: Box::new((Expr::List(body.clone()), body_span).into()),
    };
    Ok((Expr::Lambda(lambda), lambda_span).into())
}

fn eval_lambda(_: Span, lambda: &Lambda, args: &[Spanned<Expr>], env: &mut Env) -> EvalResult {
    let mut new_env = env.clone();
    let Expr::List(params) = &lambda.params.expr else {
        unreachable!("if this message is displayed, there is a bug in the in the eval function for params buts be a list");
    };
    for (param, arg) in params.iter().zip(args) {
        let Expr::Symbol(param) = &param.expr else {
            unreachable!("if this message is displayed, there is a bug in the in the eval function for params buts be a list of symbol");
        };
        new_env.data.insert(param.to_string(), eval(arg, env)?);
    }
    eval(&lambda.body, &mut new_env)
}

fn eval_if(args: &[Spanned<Expr>], env: &mut Env) -> EvalResult {
    if args.len() != 3 {
        return Err(format!(
            "invalid number of arguments for if expression expected 3 found {}: {args:#?}",
            args.len()
        ));
    }
    let Spanned {
        expr: conditional,
        span: conditional_span,
    } = eval(&args[0], env)?;

    let condition = match conditional {
        Expr::Bool(b) => b,
        _ => {
            return Err(format!(
                "if condition must evaluate to a boolean value {conditional_span:?}"
            ));
        }
    };

    let branch = if condition {
        args[1].clone()
    } else {
        args[2].clone()
    };
    eval(&branch, env)
}

fn eval_quote(args: &[Spanned<Expr>]) -> EvalResult {
    if args.len() != 1 {
        return Err("invalid number of arguments for quote expression expected 1".to_string());
    }
    Ok(args[0].clone())
}
