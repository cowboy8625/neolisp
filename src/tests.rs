use crate::environment::Env;
use crate::eval::eval;
use crate::parser::{parse_expr, parser, Builtin, Expr, Spanned};
use chumsky::prelude::*;
use pretty_assertions::assert_eq;

#[derive(Debug, Clone, PartialEq)]
pub enum TestingExpr {
    Bool(bool),
    String(String),
    Symbol(String),
    Number(f64),
    List(Vec<TestingExpr>),
    Builtin(Builtin, String),
    Func(Func),
    Lambda(Lambda),
}

impl std::fmt::Display for TestingExpr {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::Bool(b) => write!(f, "{b}"),
            Self::String(s) => write!(f, "{s}"),
            Self::Symbol(s) => write!(f, "{s}"),
            Self::Number(n) => write!(f, "{n}"),
            Self::List(list) => {
                write!(f, "(")?;
                for (i, item) in list.iter().enumerate() {
                    if i > 0 {
                        write!(f, " ")?;
                    }
                    write!(f, "{item}")?;
                }
                write!(f, ")")
            }
            Self::Builtin(b, _) => write!(f, "{b:?}"),
            Self::Func(func) => write!(f, "{func}"),
            Self::Lambda(func) => write!(f, "{func}"),
        }
    }
}

#[derive(Debug, Clone, PartialEq)]
pub struct Func {
    pub name: Box<TestingExpr>,
    pub params: Box<TestingExpr>,
    pub body: Box<TestingExpr>,
    pub help_doc: Option<Box<TestingExpr>>,
}

impl std::fmt::Display for Func {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "(fn {} {} {})", self.name, self.params, self.body)
    }
}

#[derive(Debug, Clone, PartialEq)]
pub struct Lambda {
    pub params: Box<TestingExpr>,
    pub body: Box<TestingExpr>,
}

impl std::fmt::Display for Lambda {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "(lambda {} {})", self.params, self.body)
    }
}

macro_rules! test_parser {
    (parser, $name:ident, $src:expr, $expected:expr) => {
        #[test]
        fn $name() {
            let ast = parser().parse($src).unwrap();
            let stripped = ast
                .iter()
                .cloned()
                .map(strip_out_spanned)
                .collect::<Vec<_>>();
            assert_eq!(stripped, $expected);
        }
    };
    ($name:ident, $src:expr, $expected:expr) => {
        #[test]
        fn $name() {
            let ast = parse_expr().parse($src).unwrap();
            let ast = strip_out_spanned(ast);
            assert_eq!(ast, $expected);
        }
    };
}
fn strip_out_spanned(expr: Spanned<Expr>) -> TestingExpr {
    match expr.expr {
        Expr::Bool(bool) => TestingExpr::Bool(bool),
        Expr::String(string) => TestingExpr::String(string.to_string()),
        Expr::Symbol(symbol) => TestingExpr::Symbol(symbol.to_string()),
        Expr::Number(number) => TestingExpr::Number(number),
        Expr::List(exprs) => {
            TestingExpr::List(exprs.into_iter().map(strip_out_spanned).collect::<Vec<_>>())
        }
        Expr::Builtin(builtin, help_doc) => TestingExpr::Builtin(builtin, help_doc),
        Expr::Func(func) => TestingExpr::Func(Func {
            name: Box::new(strip_out_spanned(*func.name)),
            params: Box::new(strip_out_spanned(*func.params)),
            body: Box::new(strip_out_spanned(*func.body)),
            help_doc: func.help_doc.map(|hd| Box::new(strip_out_spanned(*hd))),
        }),
        Expr::Lambda(lambda) => TestingExpr::Lambda(Lambda {
            params: Box::new(strip_out_spanned(*lambda.params)),
            body: Box::new(strip_out_spanned(*lambda.body)),
        }),
    }
}

test_parser!(
    parse_test_ident1,
    "
(hello-world)
",
    TestingExpr::List(vec![TestingExpr::Symbol("hello-world".to_string())])
);
test_parser!(
    parse_test_ident2,
    "
(*ident*)
",
    TestingExpr::List(vec![TestingExpr::Symbol("*ident*".to_string())])
);
test_parser!(
    parse_test_ident3,
    "
(<*ident/1234567890*>!?)
",
    TestingExpr::List(vec![TestingExpr::Symbol(
        "<*ident/1234567890*>!?".to_string()
    )])
);
test_parser!(
    parse_test_bool_comment,
    "
; comment
(true)
",
    TestingExpr::List(vec![TestingExpr::Bool(true)])
);
test_parser!(
    parse_test_bool_true,
    "(true)",
    TestingExpr::List(vec![TestingExpr::Bool(true)])
);
test_parser!(
    parse_test_bool_false,
    "(false)",
    TestingExpr::List(vec![TestingExpr::Bool(false)])
);
test_parser!(
    parse_test_number,
    "(1)",
    TestingExpr::List(vec![TestingExpr::Number(1.)])
);
test_parser!(
    parse_test_string,
    r#"("hello")"#,
    TestingExpr::List(vec![TestingExpr::String("hello".to_string())])
);
test_parser!(
    parse_test_string_newline,
    r#"("hello\n")"#,
    TestingExpr::List(vec![TestingExpr::String("hello\n".to_string())])
);
test_parser!(
    parse_test_string_tab,
    r#"("hello\t")"#,
    TestingExpr::List(vec![TestingExpr::String("hello\t".to_string())])
);
test_parser!(
    parse_test_symbol_plus,
    "(+ 1 2)",
    TestingExpr::List(vec![
        TestingExpr::Symbol("+".to_string()),
        TestingExpr::Number(1.),
        TestingExpr::Number(2.)
    ])
);
test_parser!(
    parse_test_symbol_minus,
    "(+ 1 2)",
    TestingExpr::List(vec![
        TestingExpr::Symbol("+".to_string()),
        TestingExpr::Number(1.),
        TestingExpr::Number(2.)
    ])
);
test_parser!(
    parse_test_symbol_gt,
    "(> 1 2)",
    TestingExpr::List(vec![
        TestingExpr::Symbol(">".to_string()),
        TestingExpr::Number(1.),
        TestingExpr::Number(2.)
    ])
);
test_parser!(
    parse_test_symbol_lt,
    "(< 1 2)",
    TestingExpr::List(vec![
        TestingExpr::Symbol("<".to_string()),
        TestingExpr::Number(1.),
        TestingExpr::Number(2.)
    ])
);
test_parser!(
    parse_test_symbol_lte,
    "(<= 1 2)",
    TestingExpr::List(vec![
        TestingExpr::Symbol("<=".to_string()),
        TestingExpr::Number(1.),
        TestingExpr::Number(2.)
    ])
);
test_parser!(
    parse_test_symbol_gte,
    "(>= 1 2)",
    TestingExpr::List(vec![
        TestingExpr::Symbol(">=".to_string()),
        TestingExpr::Number(1.),
        TestingExpr::Number(2.)
    ])
);
test_parser!(
    parse_test_list,
    "((+ 1 2) (+ 1 2))",
    TestingExpr::List(vec![
        TestingExpr::List(vec![
            TestingExpr::Symbol("+".to_string()),
            TestingExpr::Number(1.),
            TestingExpr::Number(2.)
        ]),
        TestingExpr::List(vec![
            TestingExpr::Symbol("+".to_string()),
            TestingExpr::Number(1.),
            TestingExpr::Number(2.)
        ])
    ])
);
test_parser!(
    parse_test_lambda,
    "(lambda (x y) (+ x y))",
    TestingExpr::List(vec![
        TestingExpr::Symbol("lambda".to_string()),
        TestingExpr::List(vec![
            TestingExpr::Symbol("x".to_string()),
            TestingExpr::Symbol("y".to_string())
        ]),
        TestingExpr::List(vec![
            TestingExpr::Symbol("+".to_string()),
            TestingExpr::Symbol("x".to_string()),
            TestingExpr::Symbol("y".to_string())
        ]),
    ])
);
test_parser!(
    parse_test_def_var,
    "(var x 1)",
    TestingExpr::List(vec![
        TestingExpr::Symbol("var".to_string()),
        TestingExpr::Symbol("x".to_string()),
        TestingExpr::Number(1.)
    ])
);
test_parser!(
    parse_test_def_func,
    "(fn add (x y) (+ x y))",
    TestingExpr::List(vec![
        TestingExpr::Symbol("fn".to_string()),
        TestingExpr::Symbol("add".to_string()),
        TestingExpr::List(vec![
            TestingExpr::Symbol("x".to_string()),
            TestingExpr::Symbol("y".to_string())
        ]),
        TestingExpr::List(vec![
            TestingExpr::Symbol("+".to_string()),
            TestingExpr::Symbol("x".to_string()),
            TestingExpr::Symbol("y".to_string())
        ])
    ])
);
test_parser!(
    parse_test_quote,
    "(var x '(1 2 3))",
    TestingExpr::List(vec![
        TestingExpr::Symbol("var".to_string()),
        TestingExpr::Symbol("x".to_string()),
        TestingExpr::List(vec![
            TestingExpr::Symbol("quote".to_string()),
            TestingExpr::List(vec![
                TestingExpr::Number(1.),
                TestingExpr::Number(2.),
                TestingExpr::Number(3.)
            ])
        ]),
    ])
);
test_parser!(
    parser,
    parse_test_program_with_many_single_list,
    " (x y) (+ x y) (x y z) ",
    vec![
        TestingExpr::List(vec![
            TestingExpr::Symbol("x".to_string()),
            TestingExpr::Symbol("y".to_string()),
        ]),
        TestingExpr::List(vec![
            TestingExpr::Symbol("+".to_string()),
            TestingExpr::Symbol("x".to_string()),
            TestingExpr::Symbol("y".to_string()),
        ]),
        TestingExpr::List(vec![
            TestingExpr::Symbol("x".to_string()),
            TestingExpr::Symbol("y".to_string()),
            TestingExpr::Symbol("z".to_string()),
        ])
    ]
);

// -------------------------
// |       Test Eval       |
// ------------------------

macro_rules! test_eval {
    (fail, $name:ident, $src:expr) => {
        #[test]
        fn $name() -> Result<(), String> {
            let ast = parse_expr().parse($src).map_err(|e| format!("{:?}", e))?;
            let mut env = Env::new();
            assert!(eval(&ast, &mut env).is_err());
            Ok(())
        }
    };
    (parser, $name:ident, $src:expr, $expected:expr) => {
        #[test]
        fn $name() -> Result<(), String> {
            let ast = parser().parse($src).unwrap();
            let mut env = Env::new();
            let mut result = TestingExpr::Number(0.);
            for r in ast {
                let ast = eval(&r, &mut env)?;
                result = strip_out_spanned(ast);
            }
            assert_eq!(result, $expected);
            Ok(())
        }
    };
    ($name:ident, $src:expr, $expected:expr) => {
        #[test]
        fn $name() {
            let ast = parse_expr().parse($src).unwrap();
            let mut env = Env::new();
            let r = match eval(&ast, &mut env) {
                Ok(r) => r,
                Err(e) => panic!("{:?}", e),
            };
            let result = strip_out_spanned(r);
            assert_eq!(result, $expected);
        }
    };
}
test_eval!(eval_test_symbol_add, "(+ 1 2)", TestingExpr::Number(3.));
test_eval!(eval_test_symbol_sub, "(- 1 2)", TestingExpr::Number(-1.));
test_eval!(
    eval_test_symbol_eq_false,
    "(= 1 2)",
    TestingExpr::Bool(false)
);
test_eval!(
    eval_test_symbol_eq_false_multi,
    "(= 1 2 3 4)",
    TestingExpr::Bool(false)
);
test_eval!(eval_test_symbol_eq_true, "(= 2 2)", TestingExpr::Bool(true));
test_eval!(
    eval_test_symbol_eq_true_multi,
    "(= 2 2 2 2)",
    TestingExpr::Bool(true)
);
test_eval!(
    eval_test_symbol_gt_false,
    "(> 1 2)",
    TestingExpr::Bool(false)
);
test_eval!(
    eval_test_symbol_gt_false_multi,
    "(> 1 2 3)",
    TestingExpr::Bool(false)
);
test_eval!(eval_test_symbol_gt_true, "(> 3 2)", TestingExpr::Bool(true));
test_eval!(
    eval_test_symbol_gt_true_multi,
    "(> 3 2 1)",
    TestingExpr::Bool(true)
);
test_eval!(
    eval_test_symbol_lt_false,
    "(< 2 1)",
    TestingExpr::Bool(false)
);
test_eval!(eval_test_symbol_lt_true, "(< 1 2)", TestingExpr::Bool(true));
test_eval!(
    eval_test_symbol_lt_true_multi,
    "(< 1 2 3)",
    TestingExpr::Bool(true)
);
test_eval!(
    eval_test_symbol_lt_false_multi,
    "(< 1 2 3)",
    TestingExpr::Bool(true)
);
test_eval!(
    eval_test_symbol_lte_true,
    "(<= 1 1)",
    TestingExpr::Bool(true)
);
test_eval!(
    eval_test_symbol_lte_false,
    "(<= 2 1)",
    TestingExpr::Bool(false)
);
test_eval!(
    eval_test_symbol_lte_false_multi,
    "(<= 2 3 1)",
    TestingExpr::Bool(false)
);
test_eval!(
    eval_test_symbol_gte_true,
    "(>= 1 1)",
    TestingExpr::Bool(true)
);
test_eval!(
    eval_test_symbol_gte_false,
    "(>= 1 2)",
    TestingExpr::Bool(false)
);
test_eval!(
    eval_test_symbol_and_false,
    "(and (= 1 1) (= 1 2))",
    TestingExpr::Bool(false)
);
test_eval!(
    eval_test_symbol_and_true,
    "(and (= 1 1) (<= 1 2))",
    TestingExpr::Bool(true)
);
test_eval!(
    parser,
    eval_test_fn_def_and_call,
    r#"
(fn add (x y) (+ x y))
(add 1 2)
    "#,
    TestingExpr::Number(3.)
);
test_eval!(
    parser,
    eval_test_let,
    r#"
; returns -> 3
(let
    (x 1)
    (y 2)
    (+ x y))
    "#,
    TestingExpr::Number(3.)
);
test_eval!(
    parser,
    eval_test_fn_def_var,
    r#"
(var a (+ 1 2))
    "#,
    TestingExpr::Symbol("a".to_string())
);
test_eval!(
    parser,
    eval_test_fn_def_var_with_quote,
    r#"
(var a '(+ 1 2))
(fn add (x y) a)
(add 1 2)
    "#,
    TestingExpr::List(vec![
        TestingExpr::Symbol("+".to_string(),),
        TestingExpr::Number(1.0,),
        TestingExpr::Number(2.0,),
    ])
);

// ------------------------------------------
// |       Test For Builtin Functions       |
// ------------------------------------------
test_eval!(
    eval_test_builtin_plus,
    r#"
(+ 123 321) ; -> 444
    "#,
    TestingExpr::Number(444.0)
);
test_eval!(
    eval_test_builtin_minus,
    r#"
(- 123 321) ; -> -123
    "#,
    TestingExpr::Number(-198.0)
);
test_eval!(
    eval_test_builtin_mult,
    r#"
(* 2 2) ; -> 4
    "#,
    TestingExpr::Number(4.0)
);
test_eval!(
    eval_test_builtin_div,
    r#"
(/ 100 5) ; -> 20
    "#,
    TestingExpr::Number(20.0)
);
test_eval!(
    eval_test_builtin_mod,
    r#"
(/ 100 5) ; -> 20
    "#,
    TestingExpr::Number(20.0)
);
test_eval!(
    eval_test_builtin_or,
    r#"
(or (= 1 1) (= 1 2))
    "#,
    TestingExpr::Bool(true)
);
test_eval!(
    eval_test_builtin_and,
    r#"
(and (= 1 1) (= 1 2))
    "#,
    TestingExpr::Bool(false)
);
test_eval!(
    eval_test_builtin_typeof_number,
    r#"
(typeof 1)
    "#,
    TestingExpr::Symbol("Number".to_string())
);
test_eval!(
    eval_test_builtin_typeof_string,
    r#"
(typeof "number")
    "#,
    TestingExpr::Symbol("String".to_string())
);
test_eval!(
    eval_test_builtin_typeof_list,
    r#"
(typeof (1 2 3))
    "#,
    TestingExpr::Symbol("List".to_string())
);
test_eval!(
    eval_test_builtin_list,
    r#"
(list 1 2 3)
    "#,
    TestingExpr::List(vec![
        TestingExpr::Number(1.0),
        TestingExpr::Number(2.0),
        TestingExpr::Number(3.0),
    ])
);
test_eval!(
    eval_test_builtin_cons,
    r#"
(cons (+ 0 1) (list 2 3))
    "#,
    TestingExpr::List(vec![
        TestingExpr::Number(1.0),
        TestingExpr::Number(2.0),
        TestingExpr::Number(3.0),
    ])
);
test_eval!(
    eval_test_builtin_car,
    r#"
(car (list 321 123))
    "#,
    TestingExpr::Number(321.0)
);
test_eval!(
    eval_test_builtin_cdr,
    r#"
(cdr (list 321 123))
    "#,
    TestingExpr::List(vec![TestingExpr::Number(123.0)])
);
test_eval!(
    eval_test_builtin_append,
    r#"
(append (list 1 2) (list 3 4)) ; -> (1 2 3 4)
    "#,
    TestingExpr::List(vec![
        TestingExpr::Number(1.0),
        TestingExpr::Number(2.0),
        TestingExpr::Number(3.0),
        TestingExpr::Number(4.0),
    ])
);
test_eval!(
    eval_test_builtin_reverse,
    r#"
(reverse (list 1 2 3)) ; -> (3 2 1)
    "#,
    TestingExpr::List(vec![
        TestingExpr::Number(3.0),
        TestingExpr::Number(2.0),
        TestingExpr::Number(1.0),
    ])
);
test_eval!(
    eval_test_builtin_nth,
    r#"
(nth (list 1 2 3) 2) ; -> 3
    "#,
    TestingExpr::Number(3.0)
);
test_eval!(
    eval_test_builtin_length,
    r#"
(length (list 1 2 3)) ; -> 3
    "#,
    TestingExpr::Number(3.0)
);
test_eval!(
    eval_test_builtin_map_with_lambda,
    r#"
(map (lambda (x) (+ x 1)) (list 1 2 3)) ; -> (2 3 4)
    "#,
    TestingExpr::List(vec![
        TestingExpr::Number(2.0),
        TestingExpr::Number(3.0),
        TestingExpr::Number(4.0),
    ])
);
test_eval!(
    parser,
    eval_test_builtin_map_with_function,
    r#"
(fn addone (x) (+ x 1))
(map addone (list 1 2 3)) ; -> (2 3 4)
    "#,
    TestingExpr::List(vec![
        TestingExpr::Number(2.0),
        TestingExpr::Number(3.0),
        TestingExpr::Number(4.0),
    ])
);
test_eval!(
    eval_test_builtin_fold_with_plus_op,
    r#"
(fold 0 + (list 1 2 3)) ; -> 6
    "#,
    TestingExpr::Number(6.0)
);
test_eval!(
    parser,
    eval_test_builtin_fold_with_lambda,
    r#"
; returns -> 6
(fold 0
    (lambda (x y) (+ x y))
    (list 1 2 3))
    "#,
    TestingExpr::Number(6.0)
);
test_eval!(
    parser,
    eval_test_builtin_fold_with_fn,
    r#"
(fn add (x y) (+ x y))
(fold 0 add (list 1 2 3)) ; -> 6
    "#,
    TestingExpr::Number(6.0)
);
test_eval!(
    parser,
    eval_test_builtin_filter_with_lambda,
    r#"
; returns -> (2)
(filter
    (lambda (x) (= (mod x 2) 0))
    (list 1 2 3))
    "#,
    TestingExpr::List(vec![TestingExpr::Number(2.0)])
);
test_eval!(
    parser,
    eval_test_builtin_filter_with_fn,
    r#"

(fn is-even (x) (= (mod x 2) 0))

; returns -> (2)
(filter
    is-even
    (list 1 2 3))
    "#,
    TestingExpr::List(vec![TestingExpr::Number(2.0)])
);
test_eval!(
    eval_test_builtin_loop,
    r#"(let (x 0) (loop (< x 3) (var x (+ x 1))))
; returns -> (2)
    "#,
    TestingExpr::Symbol("x".to_string())
);
test_eval!(
    fail,
    eval_test_builtin_assert,
    r#"
(assert (> 1 2) "1 is not greater than 2")
    "#
);
test_eval!(
    fail,
    eval_test_builtin_assert_eq,
    r#"
(assert 1 2 "1 is not 1 equal than 2")
    "#
);
// ----------------------------------
// |       Test For Debugging       |
// ----------------------------------

// #[test]
// fn test_debug() {
//     let src = include_str!("../test.nlisp");
//     let ast = parser().parse(src).unwrap();
//     let mut env = Env::new();
//     for node in ast {
//         eval(&node, &mut env).unwrap();
//     }
// }
