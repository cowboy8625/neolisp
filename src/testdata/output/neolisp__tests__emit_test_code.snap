---
source: src/tests.rs
expression: string
---
000000      0  jump       66
000005      5  push       1
00000F     15  push       2
000019     25  add        2
00001B     27  push       3
000025     37  eq 2
000027     39  push       <builtin 21:assert:20..40>
00003F     63  call       1
000041     65  ret-test
000042     66  push       <function 5:test()test-testing:0..42>
000066    102  set-global 0:"test()test-testing" 0..42
000086    134  get-global 0:"test()test-testing" 0..42
0000A6    166  call-test
0000A7    167  halt
